### ISSUE TITLE
Attribute names starting with ":" are not allowed anymore

### ISSUE THREAD

**Issue by lazytyper** [pugjs/pug-lexer#61]
In `vue.js` templates, HTML-attributes starting with `:` are used for passing property values to a vue component.

An example is here:

https://github.com/vuejs/vue/blob/next/examples/select2/index.html (line 27)

With `jade`, they have been accepted without any problems, but `pug` does not allow them anymore:

```
  1| 
  2| #app
> 3|   hello(:name='name')
--------------------^
  4| 

":" is not valid as the start or end of an un-quoted attribute.
at makeError (/home/egon/dev/electron-vue/node_modules/pug-error/index.js:32:13)
at Lexer.error (/home/egon/dev/electron-vue/node_modules/pug-lexer/index.js:52:15)
at Lexer.attrs (/home/egon/dev/electron-vue/node_modules/pug-lexer/index.js:1044:18)
```

**Comment by TimothyGu** [comment]
You should quote your attributes from now on, if the attribute name starts or ends with a colon.

``` jade
hello(':name'='name')
```

**Comment by lazytyper** [comment]
It does not work either, try it yourself:

```
egon@work:~/test$ pug
hello(':name'='name')
/usr/lib/node_modules/pug-cli/node_modules/pug-lexer/index.js:53
    throw err;
    ^

Error: Pug:1:16
  > 1| hello(':name'='name')
----------------------^
    2| 

":" is not valid as the start or end of an un-quoted attribute.
```

**Comment by ForbesLindesay** [comment]
In that case, that's a bug.

**Comment by bopjesvla** [comment]
Why were leading and trailing colons prohibited in the first place? As far as I can tell, browsers don't mind them.

**Comment by ForbesLindesay** [comment]
They're more difficult to lex and have very little in the way of use cases.

**Comment by aidistan** [comment]
Thousands of times using attribute names with leading colons in our app built on Vue. Better to allow it again or we have to live with the old 'jade-lexer'...

**Comment by paulpflug** [comment]
I'm also for allowing this.
but for now this works:

``` jade
someComponent(
  :name=name,
  :something=something
  )
```

just looks strange for elements with only 1 attribute.

**Comment by bopjesvla** [comment]
Plugins will enable us to transform an arbitrary shorthand to v-bind so not all is lost.

**Comment by Swimburger** [comment]
Strangely, it works for me when there's another attribute in front and when I separate them with comma's.

This works: 
`note(v-for="note in filteredNotes", :note="note" v-on:click="selectNote(note)")`

but this won't work:
`note(:note="note", v-for="note in filteredNotes"  v-on:click="selectNote(note)")`

**Comment by shunwitter** [comment]
@SNiels Thanks, It works.
I added `v-cloak` to component that needs props but have only the attribute. If you have `display: none;` on `[v-cloak]` this changes how the page is displayed before compilation, but it was ok for me.

``` pug
loading-bar(v-cloak, :progress="progress")
```

**Comment by zhyupe** [comment]
Problem is [index.js#L1071](https://github.com/pugjs/pug-lexer/blob/3ca1790b9a500eb4d1bef17e8e1b703cf4c31b5b/index.js#L1071)

Even it's quoted in templates, the key[0] is still `:`, since the quotes never been added to the key, near [index.js#L1102](https://github.com/pugjs/pug-lexer/blob/3ca1790b9a500eb4d1bef17e8e1b703cf4c31b5b/index.js#L1102).
But it's so weird that [index.js#L1075](https://github.com/pugjs/pug-lexer/blob/3ca1790b9a500eb4d1bef17e8e1b703cf4c31b5b/index.js#L1075) is trying to remove quotes.

**Comment by zhyupe** [comment]
Also, since the `trim` is executed at [index.js#L1074](https://github.com/pugjs/pug-lexer/blob/3ca1790b9a500eb4d1bef17e8e1b703cf4c31b5b/index.js#L1071), adding some white space can avoid the exception.

**Comment by zhyupe** [comment]
I wonder if anyone is working on this issue ... 

A patch for this issue can be made easily, but whether it's needed to forbid leading or trailing colon is a much more difficult problem.

This change is referred to pugjs/pug#1182, for the code below:

``` pug
option(value="level1", class : selected)
```

But the fact is: the key of an attribute should never contain white spaces, or they are two attributes, which should be separated by a comma in pug(jade).
I don't know why the developers chose to focus on colons instead of spaces, since I didn't dig much into the history of pug(jade). In my opinion, forbidding the white spaces is a better choice.

**_Update:**_
Well, I thought there should be commas ... after some googling I found it's not a requirement.

**Comment by bopjesvla** [comment]
Wait, wasn't this prohibition the result of lexing conflicts?

**Comment by bopjesvla** [comment]
If the only reason to disallow colons is "sometimes when people write colons they're making a mistake" then the check should be removed altogether.

**Comment by zhyupe** [comment]
@bopjesvla Hard to understand but it's just because someone brings other languages into pug(jade) files.

**Comment by lbineau** [comment]
Having the same issue

**Comment by wtfribley** [comment]
> In my opinion, forbidding the white spaces is a better choice.

agree. _OR_ make commas explicitly required to delimit attributes.

this problem appears when using Angular 2 syntax as well:

``` jade
li(*ngFor="p in pirates" (click)="argh()")
```

``` jade
//- works
div([class.selected]="pirate.isSelected")
div([class.selected]="parrot.isSelected" title="macaw")
//- doesn't work
div(title="macaw" [class.selected]="parrot.isSelected")
```

the second example highlights how this could be really confusing for developers, as the syntax is de-facto order-dependent. is there a danger/drawback to delimiting simply on whitespace?

at [index.js#L1042](https://github.com/pugjs/pug-lexer/blob/3ca1790b9a500eb4d1bef17e8e1b703cf4c31b5b/index.js#L1042) roughly:

``` js
return whitespaceRe.test(str[i]) || str[i] === ',';
```

**Comment by TimothyGu** [comment]
**tl;dr,** if you are using one of those frameworks that require odd\* attributes, do one of the following:
1. quote it:
   
   ``` pug
   li('*ngFor'="p in pirates" '(click)'="argh()")
   ```
2. add commas
   
   ``` pug
   li(*ngFor="p in pirates", (click)="argh()")
   ```
3. switch to a different framework that does not require the use of these specialized attribute names

---

> is there a danger/drawback to delimiting simply on whitespace?

Yes, because people (including me) like putting whitespace between operators, and even _more_ people will be wondering why their Jade templates don't work anymore after the pug@2.0.0 update.

``` pug
li(href='/item?id=' + num)
```

---

For those wondering why the Angular 2 syntax doesn't work, an explanation follows.

Pug supports virtually all JavaScript expressions in attributes. If you are using these non-standard\* attribute names, then there is a high possibility it might conflict with JavaScript syntax. For example,

``` pug
li(href='/item?id=' + num
   *ngFor=myVar)
li(href='/item?id=' + num
   (click)="argh()")
```

are semantically identical to

``` pug
li(href=(('/item?id=' + (num * ngFor)) = "p in pirates"))
li(href=(('/item?id=' + num(click)) = "argh()"))
```

There is literally no way for Pug to guess if the original intention was to multiply, or to initiate a new attribute, and any attempts at doing this would be heuristics that contain lots of inherent inconsistency and a more fundamental break of compatibility.

If you are just sticking to vanilla HTML, or even to more standard-aligning† frameworks like Polymer, then no, you will most likely not be affected by these issues at all.

Regarding attribute names contain `:` in the beginning or the ending, I _am_ having [second thoughts](https://github.com/pugjs/pug/issues/2454#issuecomment-238002983)

Since this class of issues is hard to auto-detect, the best way I see going forward, is to add a warning in our documentation against using the comma-less syntax with these frameworks.

---

I am going to lock this issue, as the subsequent discussions about other attributes were not the original issue. I am still pending @ForbesLindesay's response to possibly reverting the prohibition of attributes containing colons (the original issue), but if in a week there is still no reply I am going to revert this prohibition and make a new release. Meanwhile, to y'all developers, think about the multitude of web frameworks you are using or planning on using, and consider the necessity or benefits of a specific framework. After all, jQuery may be bad, but it's sure much simpler than all these `*ngFor` thingies. Smaller too.‡

---

*Yes, yes, strictly speaking they are _allowed_ by HTML. But before the advent of Angular 2 and Vue.js, when Pug was first developed and its syntax finalized, these syntaxes were not at all common.

†Not that I like the entire #usetheplatform campaign/fiasco, but unlike Angular, Web Components _is_ a set of standards.

‡No, I did not just endorse jQuery.
